cms/taxonomy/management/commands/sync_topics.py

```diff
@@ -0,0 +1,203 @@
+        _check_for_removed_topics({topic["id"] for topic in topics})
+
+        logger.info("Finished syncing topics.")
+
+
+def _fetch_all_topics() -> list[dict[str, str]]:
+    """Collect a complete list of topics and their subtopics by doing a
+    depth/breadth-first search using a stack of URLs.
+    """
+    topics = []
+
+    if not settings.DP_TOPIC_API_URL:
+        raise ImproperlyConfigured('"DP_TOPIC_API_URL" must be set')
+
+    # Build a stack of topics URLs and parent IDs
+    request_stack = [(f"{settings.DP_TOPIC_API_URL}/topics", None)]
+
+    # Use the stack of subtopic URls to iterate through, fetching all the subtopics
+    while request_stack:
+        url, parent_id = request_stack.pop()
+        raw_topics = _request_topics(url)
+
+        # Extract just the fields we need
+        subtopics = [
+            {
+                k: v
+                for k, v in raw_topic.items()
+                if k
+                in {
+                    "title",
+                    "id",
+                    "description",
+                }
+            }
+            for raw_topic in raw_topics
+        ]
+
+        for topic in subtopics:
+            topic["parent_id"] = parent_id
+
+        # Extend the topics list, to build a flat list of topics and subtopics which contain their own parent IDs
+        topics.extend(subtopics)
+
+        # Add any subtopics URLs from the topics we just fetched to the stack
+        request_stack.extend(_extract_subtopic_links(raw_topics))
+
+    return topics
+
+
+def _request_topics(url: str) -> list[dict[str, str]]:
+    """Fetch topics from the API and return the items from the response."""
+    topics_response = requests.get(url, timeout=30)
+    topics_response.raise_for_status()
+
+    return topics_response.json().get("items", [])
+
+
+def _extract_subtopic_links(raw_topics: list[dict[str, Any]]) -> list[tuple[str, str]]:
+    """Return a list of tuples of any subtopic links and their parent topic IDs found in raw_topics."""
+    return [
+        (subtopic_link, raw_topic["id"])
+        for raw_topic in raw_topics
+        # NOTE: This assumes we can simply call the links in the response as they are provided,
+        # perhaps switch to building the URL ourselves
+        if (
+            (subtopic_link := raw_topic.get("links", {}).get("subtopics", {}).get("href"))
+            and raw_topic.get("subtopics_ids")
+        )
+    ]
+
+
+def _sync_with_fetched_topics(fetched_topics: list[dict[str, str]]):
+    """For each fetched topic, decide if it needs to be created, updated, or left as is."""
+    _check_for_duplicate_topics(fetched_topics)
+
+    updated_count = 0
+    created_count = 0
+    for fetched_topic in fetched_topics:
+        if existing_topic := _get_topic(fetched_topic["id"]):
+            if not _topic_matches(existing_topic, fetched_topic):
+                _update_topic(existing_topic, fetched_topic)
+                updated_count += 1
```

Comment:
Log a message if they don't match?

Reply:
Could do, but the `_update_topic` function is already logging the update.